src/problems/MetalAdvectionProblem/test_sne.cpp

```diff
@@ -0,0 +1,562 @@
+{
+
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> grad_potential;
+	grad_potential[0] = 0.0;
+	grad_potential[1] = 0.0;
+
+	double z = posvec[2];
+
+	// Interpolate to find the accurate g-value from array-- because linterp doesn't work on Setonix
+	size_t i = 0;
+	double x_interp = std::abs(z);
+	amrex::GpuArray<amrex::Real, 100> xx = z_data;
+	amrex::GpuArray<amrex::Real, 100> yy = logg_data;
+	while (i < xx.size() - 1 && x_interp > xx[i + 1]) {
+		i++;
+	}
+
+	// Perform linear interpolation
+	const Real x1 = xx[i];
+	const Real x2 = xx[i + 1];
+	const Real y1 = yy[i];
+	const Real y2 = yy[i + 1];
+
+	amrex::Real ginterp = (y1 + (y2 - y1) * (x_interp - x1) / (x2 - x1));
+
+	grad_potential[2] = 2. * M_PI * Const_G * rho_dm * std::pow(R0, 2) * (2. * z / std::pow(R0, 2)) / (1. + std::pow(z, 2) / std::pow(R0, 2));
+	grad_potential[2] += 2. * M_PI * Const_G * Sigma_star * (z / z_star) * (std::pow(1. + z * z / (z_star * z_star), -0.5));
+	grad_potential[2] += (z / std::abs(z)) * std::pow(10., ginterp);
+
+	return grad_potential;
+}
+
+// Add Strang Split Source Term for External Fixed Potential Here
+template <> void QuokkaSimulation<NewProblem>::addStrangSplitSources(amrex::MultiFab &mf, int lev, amrex::Real time, amrex::Real dt_lev)
+{
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> prob_lo = geom[lev].ProbLoArray();
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> const &dx = geom[lev].CellSizeArray();
+	const Real dt = dt_lev;
+
+	for (amrex::MFIter iter(mf); iter.isValid(); ++iter) {
+		const amrex::Box &indexRange = iter.validbox();
+		auto const &state = mf.array(iter);
+
+		amrex::ParallelFor(indexRange, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+			amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> posvec, GradPhi;
+			double x1mom_new, x2mom_new, x3mom_new;
+
+			const Real rho = state(i, j, k, HydroSystem<NewProblem>::density_index);
+			const Real x1mom = state(i, j, k, HydroSystem<NewProblem>::x1Momentum_index);
+			const Real x2mom = state(i, j, k, HydroSystem<NewProblem>::x2Momentum_index);
+			const Real x3mom = state(i, j, k, HydroSystem<NewProblem>::x3Momentum_index);
+			const Real Egas = state(i, j, k, HydroSystem<NewProblem>::energy_index);
+
+			Real Eint = RadSystem<NewProblem>::ComputeEintFromEgas(rho, x1mom, x2mom, x3mom, Egas);
+
+			posvec[0] = prob_lo[0] + (i + 0.5) * dx[0];
+
+#if (AMREX_SPACEDIM >= 2)
+			posvec[1] = prob_lo[1] + (j + 0.5) * dx[1];
+#endif
+
+#if (AMREX_SPACEDIM >= 3)
+			posvec[2] = prob_lo[2] + (k + 0.5) * dx[2];
+#endif
+
+			GradPhi = HydroSystem<NewProblem>::GetGradFixedPotential(posvec);
+
+			x1mom_new = state(i, j, k, HydroSystem<NewProblem>::x1Momentum_index) + dt * (-rho * GradPhi[0]);
+			x2mom_new = state(i, j, k, HydroSystem<NewProblem>::x2Momentum_index) + dt * (-rho * GradPhi[1]);
+			x3mom_new = state(i, j, k, HydroSystem<NewProblem>::x3Momentum_index) + dt * (-rho * GradPhi[2]);
```

Comment:
`state(i, j, k, HydroSystem<NewProblem>::x3Momentum_index)` can be replaced with `x1mom`. The whole string can be replaced here, not just the component index.

Reply:
I am a bit confused here. Because previously, it was - 
`x1mom_new = state(i, j, k, HydroSystem<NewProblem>::x1Momentum_index) + dt * (-rho * GradPhi[0]);`

and then - 
`state(i, j, k, x1mom) = x1mom_new`

-but this was the bit that was causing problem. For some reason, the density was turning negative between these sets of code. So, I switched it to `state(i, j, k, HydroSystem<NewProblem>::x1Momentum_index)` and it worked fine.

Reply:
This works:
```
const double x1mom_new = x1mom + dt * (-rho * GradPhi[0]);
state(i, j, k, HydroSystem<NewProblem>::x1Momentum_index) = x1mom_new;
```